### PR TITLE
added user query for keys when adding to an associative container

### PR DESCRIPTION
**Pull request by alexmontAmazon**
Signed-off-by: Alex Montgomery <alexmont@amazon.com>

## What does this PR do?
Adds a mechanism for the DPE to query keys from the user, whenever a new entry is added to an associative container

Note: Although this is functional, it has some rough edges to address in a future PR, possibly by someone other than me:
1) Adapter shows a superfluous root node in the DPE embedded in the key query dialog. This should be suppressed using the same mechanism as AZ::Edit::PropertyVisibility::ShowChildrenOnly.
2) When creating integral types (likesay, an integer) the container does not zero construct the key, so the key query dialog will suggest something like, "-2556236", which isn't a great experience.
3) When triggering the key query, the container reserves a new entry before passing it via the adapter to the DPE. If the user cancels out of the key query, that entry is allocated but never stored. I'm honestly not sure if that's a leak or if that's okay as-is.

## How was this PR tested?
locally in DPEStandAlone

image

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp [hunk 1]
```diff
@@ -92,7 +92,28 @@ namespace AZ::DocumentPropertyEditor
...
                     m_container->GetElement(m_container->GetDefaultElementNameCrc());
+
+                // dataAddress is allocated instance of key type. Get the key type and send it with the dataAddress in the
+                // message, then skip the store element below until we get an AddContainerKey message back from the DPE UI
                 void* dataAddress = m_container->ReserveElement(m_instance, containerClassElement);
+
+                auto associativeContainer = m_container->GetAssociativeContainerInterface();
+                if (associativeContainer)
+                {
+                    auto keyTypeAttribute = containerClassElement->FindAttribute(AZ_CRC("KeyType"));
```
Comment (cgalvan) [inline Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:103]: Nit: Should we use `AZ_CRC_CE` here so the compiler can compute the hash for us?

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 2]
```diff
@@ -902,7 +910,7 @@ namespace AzToolsFramework
         const bool indexInRange = (rowIndex <= m_domOrderedRows.size());
         AZ_Assert(indexInRange, "rowIndex cannot be more than one past the existing end!")
 
-        if (indexInRange)
+            if (indexInRange)
```
Comment (cgalvan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:913]: Nit: Looks like some extra whitespace here

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 3]
```diff
@@ -1023,6 +1031,24 @@ namespace AzToolsFramework
         }
     }
 
+    void DocumentPropertyEditor::HandleDomMessage(
+        const AZ::DocumentPropertyEditor::AdapterMessage& message, [[maybe_unused]] AZ::Dom::Value& value)
+    {
+        // message match for QueryKey
+        auto showKeyQueryDialog = [&](AZ::DocumentPropertyEditor::DocumentAdapterPtr* adapter, AZ::Dom::Path containerPath)
+        {
+            QDialog keyQueryDialog;
```
Comment (cgalvan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:1040]: Is this `QDialog` unused?

**Review by cgalvan — APPROVED** [review]

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.h [hunk 4]
```diff
@@ -60,6 +60,9 @@ namespace AZ::DocumentPropertyEditor
         //! property editor instances has altered its value.
         void NotifyPropertyChanged(const PropertyChangeInfo& changeInfo);
 
+        void* GetInstance() { return m_instance; }
+        AZ::TypeId GetTypeId() { return m_typeId; }
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.h:64]: The GetTypeId function could be const.
There can also be as second overload of `GetInstance()` that returns a `const void*` that is also a const member.

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp [hunk 5]
```diff
@@ -583,6 +609,18 @@ namespace AZ::DocumentPropertyEditor
                 }
             }
         };
+        auto addKeyToContainer = [&](AZ::DocumentPropertyEditor::DocumentAdapterPtr* adapter, AZ::Dom::Path containerPath)
+        {
+            ReflectionAdapter* actualAdapter = static_cast<ReflectionAdapter*>(adapter->get());
+            auto containerEntry = m_impl->m_containers.ValueAtPath(containerPath, AZ::Dom::PrefixTreeMatch::ParentsOnly);
+            void* keyInstance = actualAdapter->GetInstance();
+
+            containerEntry->m_container->StoreElement(containerEntry->m_instance, keyInstance);
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:618]: If this addKeyToContainer function isn't called, then this is leaking memory.
We should add logic for when the Add Key dialog is not accepted to invoke RemoveElement on the container.

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:103]: Yes we should.
Or if there is a constant, we can use that as well

File: Code/Framework/AzCore/AzCore/DOM/DomPath.h [hunk 6]
```diff
@@ -68,6 +68,8 @@ namespace AZ::Dom
     class Path final
     {
     public:
+        AZ_TYPE_INFO(Path, "{C0081C45-F15D-4F46-9680-19535D33C312}")
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/DOM/DomPath.h:71]: Could you add the namespace of `AZ::Dom::Path` here?
We actually have TypeInfo for [AZ::IO::Path](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/IO/Path/Path.h#L740-L744) and using the fully qualified namespace would help disambiguate

**Review by lumberyard-employee-dm — COMMENTED** [review]
The AssociativeContainer Interface has an [AllocateKey](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl#L805-L812) function, which initialize Standard layout types such int to 0.

So that can be used instead of ReserveElement.

However, I also think we should update the [AZStdAssociativeContainer::ReserveElement](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/Serialization/AZStdContainers.inl#L796-L801) function to default construct, the ValueType

That is as simple as adding a curly brace to the end of line 800 on the ValueType construction

`return new(containerPtr->get_allocator().allocate(sizeof(ValueType), AZStd::alignment_of<ValueType>::value))ValueType{};`

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp [hunk 7]
```diff
@@ -529,6 +554,8 @@ namespace AZ::DocumentPropertyEditor
     Dom::Value ReflectionAdapter::GenerateContents()
     {
         m_impl->m_builder.BeginAdapter();
+        m_impl->m_builder.AddMessageHandler(this, Nodes::Adapter::QueryKey);
+        m_impl->m_builder.AddMessageHandler(this, Nodes::Adapter::AddContainerKey);
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:558]: How about adding a `Node::Adapter::RejectContainerKey` to handle the case where a key has not been accepted into the DOM?

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp (hunk 5 above)
Comment (alexmontAmazon) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:618]: awesome, thanks. I wasn't clear on that, as I put in the note :)

File: Code/Framework/AzCore/AzCore/DOM/DomPath.h (hunk 6 above)
Comment (alexmontAmazon) [reply Code/Framework/AzCore/AzCore/DOM/DomPath.h:71]: sure, will do

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp (hunk 7 above)
Comment (alexmontAmazon) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:558]: yep! Good plan :)

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp (hunk 1 above)
Comment (alexmontAmazon) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.cpp:103]: sure, will do 👍

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 3 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:1040]: yikes! That was from testing. Good catch!

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.h (hunk 4 above)
Comment (alexmontAmazon) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/ReflectionAdapter.h]: done.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 2 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp]: fixed, thanks!

**Review by lumberyard-employee-dm — APPROVED** [review]

**Comment by lumberyard-employee-dm** [comment]
> Signed-off-by: Alex Montgomery [alexmont@amazon.com](mailto:alexmont@amazon.com)
> ## What does this PR do?
> 
> Adds a mechanism for the DPE to query keys from the user, whenever a new entry is added to an associative container
> 
> Note: Although this is functional, it has some rough edges to address in a future PR, possibly by someone other than me:
> 
>     1. Adapter shows a superfluous root node in the DPE embedded in the key query dialog. This should be suppressed using the same mechanism as AZ::Edit::PropertyVisibility::ShowChildrenOnly.
> 
>     2. When creating integral types (likesay, an integer) the container does not zero construct the key, so the key query dialog will suggest something like, "-2556236", which isn't a great experience.
> 
>     3. When triggering the key query, the container reserves a new entry before passing it via the adapter to the DPE. If the user cancels out of the key query, that entry is allocated but never stored. I'm honestly not sure if that's a leak or if that's okay as-is.
> 
> 
> ## How was this PR tested?
> 
> locally in DPEStandAlone
> 
> image

Looks like you can remove issues number 2 and 3 from the PR description now.

**Review by amzn-tmryan — APPROVED** [review]

**Pull request merged**